engine/src/main/java/org/pentaho/di/trans/steps/dynamicsqlrow/DynamicSQLRow.java

```diff
@@ -0,0 +1,343 @@
+          newRow[newIndex++] = data.notfound[i];
+        }
+        putRow( data.outputRowMeta, newRow );
+
+        if ( meta.isQueryOnlyOnChange() ) {
+          // add row to the previous rows buffer
+          data.previousrowbuffer.add( data.notfound );
+          data.skipPreviousRow = false;
+        }
+      } else {
+        if ( meta.isQueryOnlyOnChange() && counter == 0 && !meta.isOuterJoin() ) {
+          data.skipPreviousRow = true;
+        }
+      }
+
+      if ( data.db != null ) {
+        data.db.closeQuery( rs );
+      }
+    }
+
+  }
+
+  public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
+    meta = (DynamicSQLRowMeta) smi;
+    data = (DynamicSQLRowData) sdi;
+
+    Object[] r = getRow(); // Get row from input rowset & set row busy!
+    if ( r == null ) { // no more input to be expected...
+      setOutputDone();
+      return false;
+    }
+    if ( first ) {
+      if ( Utils.isEmpty( meta.getSQLFieldName() ) ) {
+        throw new KettleException( BaseMessages.getString( PKG, "DynamicSQLRow.Exception.SQLFieldNameEmpty" ) );
+      }
+
+      if ( Utils.isEmpty( meta.getSql() ) ) {
+        throw new KettleException( BaseMessages.getString( PKG, "DynamicSQLRow.Exception.SQLEmpty" ) );
+      }
+
+      // cache the position of the field
+      if ( data.indexOfSQLField < 0 ) {
+        data.indexOfSQLField = getInputRowMeta().indexOfValue( meta.getSQLFieldName() );
+        if ( data.indexOfSQLField < 0 ) {
+          // The field is unreachable !
+          throw new KettleException( BaseMessages.getString( PKG, "DynamicSQLRow.Exception.FieldNotFound", meta
+            .getSQLFieldName() ) );
+        }
+      }
+    }
+    try {
+      lookupValues( getInputRowMeta(), r );
+
+      if ( checkFeedback( getLinesRead() ) ) {
+        if ( log.isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "DynamicSQLRow.Log.LineNumber" ) + getLinesRead() );
+        }
+      }
+    } catch ( KettleException e ) {
+      boolean sendToErrorRow = false;
+      String errorMessage = null;
+
+      if ( getStepMeta().isDoingErrorHandling() ) {
+        sendToErrorRow = true;
+        errorMessage = e.toString();
+      } else {
+        logError( BaseMessages.getString( PKG, "DynamicSQLRow.Log.ErrorInStepRunning" ) + e.getMessage() );
+        setErrors( 1 );
+        stopAll();
+        setOutputDone(); // signal end to receiver(s)
+        return false;
+      }
+      if ( sendToErrorRow ) {
+        // Simply add this row to the error row
+        putError( getInputRowMeta(), r, 1, errorMessage, null, "DynamicSQLRow001" );
+      }
+    }
+
+    return true;
+  }
+
+  /** Stop the running query */
+  public void stopRunning( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
+    meta = (DynamicSQLRowMeta) smi;
+    data = (DynamicSQLRowData) sdi;
+
+    if ( data.db != null && !data.isCanceled ) {
+      synchronized ( data.db ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object data.db synchronized on line 285 should not be re-assigned in the same synchronized block on line 286. After this re-assignment, you lose the reference to the synchronized object.